### PR TITLE
Feature/app 2819 query supported currencies

### PR DESCRIPTION
**Pull request by Gracian0-Caldeir4**
**What does this PR do?**

   This PR adds a new service to obtain a list of supported currencies. Each element of this list will be represent by Currency new entity. It will also add a CurrencyUseCases and a CurrencyRepository.

**Database changed?**

   No

**Where should the reviewer start?**

- [ ] Currency.swift
- [ ] CurrencyUseCases.swift
- [ ] CurrencyRepository.swift
- [ ] AppCoinsBillingClient.swift

**How should this be manually tested?**

  Try to call the new request called getSupportedCurrencies() and verify if it is returning a list of Currency elements. Also you can make the request with this url ("(endpoint)/currencies?type=FIAT&icon.height=128&limit=10") and verify if the method is returning the same list.

**What are the relevant tickets?**

  Tickets related to this pull-request: APP-2819

**Questions:**

   Does this add new dependencies which need to be added? (Eg. new keys, etc.) 

**Code Review Checklist**

- [ ] Documentation on public interfaces
- [ ] Database changed? If yes - Migration?
- [ ] Remove comments & unused code & forgotten testing Logs
- [ ] Codestyle
- [ ] New Kotlin code has unit tests
- [ ] New flows in presenters unit tests
- [ ] Mappers/Validators with any kind of logic unit tests
- [ ] Functional tests pass

### CODE REVIEW

File: Sources/AppCoinsSDK/Domain/Entities/Currency.swift [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+    let sign: String
+    let type: String
+    let flag: URL?
+    
+    init(currencyRaw: CurrencyRaw) {
+        self.currency = currencyRaw.currency
+        self.label = currencyRaw.label
+        self.sign = currencyRaw.sign
+        self.type = currencyRaw.type
+        if let url = URL(string: currencyRaw.flag) {
```
Comment (tiaguu) [inline Sources/AppCoinsSDK/Domain/Entities/Currency.swift:23]: `self.url = URL(string: currencyRaw.flag)`

Will do the trick, no? No need for 3 lines

File: Sources/AppCoinsSDK/Domain/Repositories/Currency/CurrencyRepository.swift [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+            completion(.success(currencyList))
+        } else {
+            self.AppCoinBillingService.getSupportedCurrencies { result in
+                switch result {
+                case .success(let currencyListRaw):
+                    var currencyList: [Currency] = []
+                    for itemRaw in currencyListRaw.items {
+                        currencyList.append(Currency(currencyRaw: itemRaw))
+                    }
+                    self.CurrencyListCache.setValue(currencyList, forKey: "currencyList", storageOption: .disk(ttl: 24 * 3600))
```
Comment (tiaguu) [inline Sources/AppCoinsSDK/Domain/Repositories/Currency/CurrencyRepository.swift:26]: We should cache this in the disk for 30 days, so this should be `ttl: 30 * 24 * 3600`

File: Sources/AppCoinsSDK/Models/Raws/CurrencyListRaw.swift [hunk 3]
```diff
@@ -0,0 +1,20 @@
+//
+//  File.swift
+//  
+//
+//  Created by Graciano Caldeira on 16/08/2024.
+//
+
+import Foundation
+
+internal struct CurrencyListRaw: Codable {
```
Comment (tiaguu) [inline Sources/AppCoinsSDK/Models/Raws/CurrencyListRaw.swift:10]: No need for separate files, add this to the CurrencyRaw file

File: Sources/AppCoinsSDK/Models/Services/Billing/AppCoinBillingClient.swift [hunk 4]
```diff
@@ -553,4 +553,32 @@ internal class AppCoinBillingClient : AppCoinBillingService {
         }
     }
     
+    internal func getSupportedCurrencies(result: @escaping (Result<CurrencyListRaw, BillingError>) -> Void) {
```
Comment (tiaguu) [inline Sources/AppCoinsSDK/Models/Services/Billing/AppCoinBillingClient.swift:556]: We should deal with pagination. The request returns `next` and `previous` pointers, so we should query the rest of the currencies if next is not `null`.

**Review by tiaguu — CHANGES_REQUESTED** [review]

File: Sources/AppCoinsSDK/Domain/Entities/Currency.swift (hunk 1 above)
Comment (Gracian0-Caldeir4) [reply Sources/AppCoinsSDK/Domain/Entities/Currency.swift]: Done

File: Sources/AppCoinsSDK/Domain/Repositories/Currency/CurrencyRepository.swift (hunk 2 above)
Comment (Gracian0-Caldeir4) [reply Sources/AppCoinsSDK/Domain/Repositories/Currency/CurrencyRepository.swift]: Done

File: Sources/AppCoinsSDK/Models/Raws/CurrencyListRaw.swift (hunk 3 above)
Comment (Gracian0-Caldeir4) [reply Sources/AppCoinsSDK/Models/Raws/CurrencyListRaw.swift]: Done

File: Sources/AppCoinsSDK/Models/Services/Billing/AppCoinBillingClient.swift (hunk 4 above)
Comment (Gracian0-Caldeir4) [reply Sources/AppCoinsSDK/Models/Services/Billing/AppCoinBillingClient.swift]: Done

File: Sources/AppCoinsSDK/Models/Services/Billing/AppCoinBillingClient.swift [hunk 5]
```diff
@@ -553,32 +553,44 @@ internal class AppCoinBillingClient : AppCoinBillingService {
...
-            
-            request.httpMethod = "GET"
-            request.addValue("application/json", forHTTPHeaderField: "Content-Type")
-            request.timeoutInterval = 10
-            
-            let task = URLSession.shared.dataTask(with: request) { data, response, error in
+    internal func getSupportedCurrencies(result: @escaping (Result<[CurrencyRaw], BillingError>) -> Void) {
+        var allCurrencies: [CurrencyRaw] = []
+        
+        func getAllCurrencies(url: URL? = nil) {
```
Comment (tiaguu) [inline Sources/AppCoinsSDK/Models/Services/Billing/AppCoinBillingClient.swift:559]: I really like this solution. Well done. 

Just three tiny details I would improve:
1. I would build the first URL outside the getAllCurrencies() method and make the url parameter non-optional.
2. Call the `allCurrencies` variable simply `currencies` – no need to add complexity to the variable names.
3. Call the `getAllCurrencies` method `getNextCurrenciesBatch` – this is picky but what we're doing here is recursively getting batches of currencies, not getting them all as the method name indicates.

These are all really small details and are suggestions. The solution is OK and I will not reject the PR if you decide to move forward like this.

File: Sources/AppCoinsSDK/Models/Services/Billing/AppCoinBillingClient.swift [hunk 6]
```diff
@@ -553,32 +553,44 @@ internal class AppCoinBillingClient : AppCoinBillingService {
...
+            if let requestURL = url ?? URL(string: "\(endpoint)/currencies?type=FIAT&icon.height=128") {
```
Comment (tiaguu) [inline Sources/AppCoinsSDK/Models/Services/Billing/AppCoinBillingClient.swift:560]: When building the URL use `URLComponents` as done on the methods at the top of this class.

**Review by tiaguu — COMMENTED** [review]
Everything looks fine. I really like the solution for the getSupportedCurrencies() recursive method, I just added some small suggestions that you might want to consider.

**Review by tiaguu — APPROVED** [review]

**Pull request merged**